File: ghrah-protocol/src/ghrah/protocol/payloads/session.py
# SPDX-FileCopyrightText: 2026 chenxya <chenxya@ghrah.org>
#
# SPDX-License-Identifier: Apache-2.0
"""Session/Branch 域载荷模型（命令、事件与链历史查询）。"""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field

# ─── Session 命令载荷模型 ───


class SessionCreatePayload(BaseModel):
    """session_create 命令载荷。"""

    project_id: str
    agent_id: str
    agent_name: str
    origin_session_id: str | None = None
    origin_node_id: str | None = None
    system_prompt: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class SessionActivatePayload(BaseModel):
    """session_activate 命令载荷。"""

    project_id: str
    agent_id: str
    agent_name: str
    session_id: str


class SessionListPayload(BaseModel):
    """session_list 命令载荷。"""

    project_id: str
    agent_id: str
    agent_name: str


class SessionArchivePayload(BaseModel):
    """session_archive 命令载荷。"""

    project_id: str
    agent_id: str
    agent_name: str
    session_id: str


class SessionDeletePayload(BaseModel):
    """session_delete 命令载荷。"""

    project_id: str
    agent_id: str
    agent_name: str
    session_id: str


class BranchCreatePayload(BaseModel):
    """branch_create 命令载荷。"""

    project_id: str
    agent_id: str
    agent_name: str
    session_id: str
    name: str
    from_node_id: str | None = None
    parent_branch_id: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class BranchActivatePayload(BaseModel):
    """branch_activate 命令载荷。"""

    project_id: str
    agent_id: str
    agent_name: str
    session_id: str
    branch_id: str


class BranchListPayload(BaseModel):
    """branch_list 命令载荷。"""

    project_id: str
    agent_id: str
    agent_name: str
    session_id: str


class BranchArchivePayload(BranchActivatePayload):
    """branch_archive 命令载荷。"""


class BranchDeletePayload(BranchActivatePayload):
    """branch_delete 命令载荷。"""


class GetChainHistoryPayload(BaseModel):
    """get_chain_history 命令载荷。"""

    project_id: str
    agent_id: str
    agent_name: str
    session_id: str
    branch_id: str
    limit: int = -1


class ChainHistoryResultPayload(BaseModel):
    """get_chain_history 命令的响应载荷。"""

    project_id: str
    agent_id: str
    agent_name: str
    session_id: str
    branch_id: str
    active_session_id: str = ""
    nodes: list[dict[str, Any]] = Field(default_factory=list)  # ─── Session 事件载荷模型 ───


class SessionInfoPayload(BaseModel):
    """Session 信息载荷，用于 session 事件和查询结果。"""

    project_id: str = ""
    agent_id: str = ""
    cluster_id: str = ""
    session_id: str
    agent_name: str
    root_node_id: str
    active_branch_id: str
    state: str = "active"
    system_prompt: str = ""
    origin_session_id: str | None = None
    origin_node_id: str | None = None
    created_at: str = ""
    metadata: dict[str, Any] = Field(default_factory=dict)
    message_count: int = 0
    iteration_count: int = 0


class SessionCreatedPayload(BaseModel):
    """session_created 事件载荷。"""

    project_id: str = ""
    agent_id: str = ""
    cluster_id: str = ""
    agent_name: str
    session: SessionInfoPayload


class SessionActivatedPayload(BaseModel):
    """session_activated 事件载荷。"""

    project_id: str = ""
    agent_id: str = ""
    cluster_id: str = ""
    agent_name: str
    session: SessionInfoPayload


class SessionArchivedPayload(BaseModel):
    """session_archived 事件载荷。"""

    project_id: str = ""
    agent_id: str = ""
    cluster_id: str = ""
    agent_name: str
    session_id: str


class SessionDeletedPayload(BaseModel):
    """session_deleted 事件载荷。"""

    project_id: str = ""
    agent_id: str = ""
    cluster_id: str = ""
    agent_name: str
    session_id: str


class SessionListResultPayload(BaseModel):
    """session_list_result 事件载荷。"""

    project_id: str = ""
    agent_id: str = ""
    cluster_id: str = ""
    agent_name: str
    sessions: list[SessionInfoPayload]


class BranchInfoPayload(BaseModel):
    """Branch 信息载荷。"""

    branch_id: str
    session_id: str
    name: str
    head_node_id: str
    parent_branch_id: str | None = None
    fork_point_node_id: str | None = None
    created_at: str = ""
    metadata: dict[str, Any] = Field(default_factory=dict)


class BranchEventPayload(BaseModel):
    """Branch 创建/激活事件载荷。"""

    project_id: str = ""
    agent_id: str = ""
    cluster_id: str = ""
    agent_name: str
    branch: BranchInfoPayload


class BranchLifecyclePayload(BaseModel):
    """Branch 归档/删除事件载荷。"""

    project_id: str = ""
    agent_id: str = ""
    cluster_id: str = ""
    agent_name: str
    session_id: str
    branch_id: str


class BranchListResultPayload(BaseModel):
    """branch_list_result 事件载荷。"""

    project_id: str = ""
    agent_id: str = ""
    cluster_id: str = ""
    agent_name: str
    session_id: str
    branches: list[BranchInfoPayload]

File: ghrah-observer-webui/packages/protocol/src/payloads/session.ts
import { z } from "zod";

// ─── Agent-scoped Session / Chain Payloads ───

export const SessionCreatePayloadSchema = z.object({
  project_id: z.string(),
  agent_id: z.string(),
  agent_name: z.string(),
  origin_session_id: z.string().nullable().optional(),
  origin_node_id: z.string().nullable().optional(),
  system_prompt: z.string().nullable().optional(),
  metadata: z.record(z.unknown()).optional().default({}),
});

export const SessionActivatePayloadSchema = z.object({
  project_id: z.string(),
  agent_id: z.string(),
  agent_name: z.string(),
  session_id: z.string(),
});

export const SessionListPayloadSchema = z.object({
  project_id: z.string(),
  agent_id: z.string(),
  agent_name: z.string(),
});

export const SessionArchivePayloadSchema = SessionActivatePayloadSchema;
export const SessionDeletePayloadSchema = SessionActivatePayloadSchema;

export const BranchCreatePayloadSchema = z.object({
  project_id: z.string(),
  agent_id: z.string(),
  agent_name: z.string(),
  session_id: z.string(),
  name: z.string(),
  from_node_id: z.string().nullable().optional(),
  parent_branch_id: z.string().nullable().optional(),
  metadata: z.record(z.unknown()).optional().default({}),
});

export const BranchActivatePayloadSchema = z.object({
  project_id: z.string(),
  agent_id: z.string(),
  agent_name: z.string(),
  session_id: z.string(),
  branch_id: z.string(),
});

export const BranchListPayloadSchema = BranchActivatePayloadSchema.omit({ branch_id: true });
export const BranchArchivePayloadSchema = BranchActivatePayloadSchema;
export const BranchDeletePayloadSchema = BranchActivatePayloadSchema;

export const GetChainHistoryPayloadSchema = z.object({
  project_id: z.string(),
  agent_id: z.string(),
  agent_name: z.string(),
  session_id: z.string(),
  branch_id: z.string(),
  limit: z.number().int().optional().default(-1),
});

export const ChainHistoryResultPayloadSchema = z.object({
  project_id: z.string(),
  agent_id: z.string(),
  agent_name: z.string(),
  session_id: z.string(),
  branch_id: z.string(),
  active_session_id: z.string().optional().default(""),
  nodes: z.array(z.record(z.unknown())).optional().default([]),
});

export const SessionInfoPayloadSchema = z.object({
  project_id: z.string().optional().default(""),
  agent_id: z.string().optional().default(""),
  cluster_id: z.string().optional().default(""),
  session_id: z.string(),
  agent_name: z.string(),
  root_node_id: z.string(),
  active_branch_id: z.string(),
  state: z.string().optional().default("active"),
  system_prompt: z.string().optional().default(""),
  origin_session_id: z.string().nullable().optional(),
  origin_node_id: z.string().nullable().optional(),
  created_at: z.string().optional().default(""),
  metadata: z.record(z.unknown()).optional().default({}),
  message_count: z.number().int().optional().default(0),
  iteration_count: z.number().int().optional().default(0),
});

export const SessionCreatedPayloadSchema = z.object({
  project_id: z.string().optional().default(""),
  agent_id: z.string().optional().default(""),
  cluster_id: z.string().optional().default(""),
  agent_name: z.string(),
  session: SessionInfoPayloadSchema,
});

export const SessionActivatedPayloadSchema = SessionCreatedPayloadSchema;

export const SessionArchivedPayloadSchema = z.object({
  project_id: z.string().optional().default(""),
  agent_id: z.string().optional().default(""),
  cluster_id: z.string().optional().default(""),
  agent_name: z.string(),
  session_id: z.string(),
});

export const SessionDeletedPayloadSchema = SessionArchivedPayloadSchema;

export const SessionListResultPayloadSchema = z.object({
  project_id: z.string().optional().default(""),
  agent_id: z.string().optional().default(""),
  cluster_id: z.string().optional().default(""),
  agent_name: z.string(),
  sessions: z.array(SessionInfoPayloadSchema),
});

export const BranchInfoPayloadSchema = z.object({
  branch_id: z.string(),
  session_id: z.string(),
  name: z.string(),
  head_node_id: z.string(),
  parent_branch_id: z.string().nullable().optional(),
  fork_point_node_id: z.string().nullable().optional(),
  created_at: z.string().optional().default(""),
  metadata: z.record(z.unknown()).optional().default({}),
});

export const BranchEventPayloadSchema = z.object({
  project_id: z.string().optional().default(""),
  agent_id: z.string().optional().default(""),
  cluster_id: z.string().optional().default(""),
  agent_name: z.string(),
  branch: BranchInfoPayloadSchema,
});

export const BranchLifecyclePayloadSchema = z.object({
  project_id: z.string().optional().default(""),
  agent_id: z.string().optional().default(""),
  cluster_id: z.string().optional().default(""),
  agent_name: z.string(),
  session_id: z.string(),
  branch_id: z.string(),
});

export const BranchListResultPayloadSchema = z.object({
  project_id: z.string().optional().default(""),
  agent_id: z.string().optional().default(""),
  cluster_id: z.string().optional().default(""),
  agent_name: z.string(),
  session_id: z.string(),
  branches: z.array(BranchInfoPayloadSchema),
});

export type SessionCreatePayload = z.infer<typeof SessionCreatePayloadSchema>;
export type SessionActivatePayload = z.infer<typeof SessionActivatePayloadSchema>;
export type SessionListPayload = z.infer<typeof SessionListPayloadSchema>;
export type SessionArchivePayload = z.infer<typeof SessionArchivePayloadSchema>;
export type SessionDeletePayload = z.infer<typeof SessionDeletePayloadSchema>;
export type GetChainHistoryPayload = z.infer<typeof GetChainHistoryPayloadSchema>;
export type ChainHistoryResultPayload = z.infer<typeof ChainHistoryResultPayloadSchema>;
export type SessionInfoPayload = z.infer<typeof SessionInfoPayloadSchema>;
export type SessionCreatedPayload = z.infer<typeof SessionCreatedPayloadSchema>;
export type SessionActivatedPayload = z.infer<typeof SessionActivatedPayloadSchema>;
export type SessionArchivedPayload = z.infer<typeof SessionArchivedPayloadSchema>;
export type SessionDeletedPayload = z.infer<typeof SessionDeletedPayloadSchema>;
export type SessionListResultPayload = z.infer<typeof SessionListResultPayloadSchema>;
export type BranchCreatePayload = z.infer<typeof BranchCreatePayloadSchema>;
export type BranchActivatePayload = z.infer<typeof BranchActivatePayloadSchema>;
export type BranchListPayload = z.infer<typeof BranchListPayloadSchema>;
export type BranchArchivePayload = z.infer<typeof BranchArchivePayloadSchema>;
export type BranchDeletePayload = z.infer<typeof BranchDeletePayloadSchema>;
export type BranchInfoPayload = z.infer<typeof BranchInfoPayloadSchema>;
export type BranchEventPayload = z.infer<typeof BranchEventPayloadSchema>;
export type BranchLifecyclePayload = z.infer<typeof BranchLifecyclePayloadSchema>;
export type BranchListResultPayload = z.infer<typeof BranchListResultPayloadSchema>;
